core/src/main/scala/com/salesforce/op/stages/impl/tuning/OpValidator.scala

```diff
@@ -287,25 +321,41 @@ private[op] trait OpValidator[M <: Model[_], E <: Estimator[_]] extends Serializ
           estimator.set(pi1, label).set(pi2, features)
       }
       Future {
-        val numModels = params.length
-        val metrics = new Array[Double](params.length)
-        log.info(s"Train split with multiple sets of parameters.")
-        val models = estimator.fit(train, params).asInstanceOf[Seq[M]]
-        for {i <- 0 until numModels} {
-          val metric = evaluator.evaluate(models(i).transform(test, params(i)))
-          log.info(s"Got metric $metric for model $name trained with ${params(i)}.")
-          metrics(i) = metric
+        val paramsMetrics = params.map { p =>
+          Try {
```

Comment:
You might wanna consider to parallelize by splits AND models, I.e just use futures instead of tries nested inside futures.

Reply:
I tried that but I think the copy of the model causes some kind of lock - when I had them both in futures the first test where the model selectors are ran for 30 minutes without ever finishing....

Reply:
Let’s put a todo there or perhaps Chris S. Can help?

Reply:
@tovbinm @leahmcguire I was able to replace `params` with `params.par` and although it worked it was much slower than one would expect. First, does it make sense to fit and eval all models in parallel when there is only one spark context? 
BTW do you have `jvisualvm` avail in Zulu?

Reply:
You can always get visualvm standalone https://visualvm.github.io/download.html

Reply:
https://github.com/salesforce/TransmogrifAI/pull/373 - work fine for me. Though I am not sure about running anything in parallel with spark context as dependency.

Reply:
Please refer to the implementation of cross validation in Spark (it uses Futures and there is a good reason why).

Reply:
And that’s why I propose to stay with Futures, but just use them correctly.

Reply:
@tovbinm If you are referring to https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala then the only difference with my PR is that they are using `foldMetricFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))` instead of `Future.sequence` therefore can you elaborate more?